File: library/src/main/java/com/github/ohohcakester/algorithms/astar/BreadthFirstSearch.java
package com.github.ohohcakester.algorithms.astar;

import com.github.ohohcakester.grid.GridGraph;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class BreadthFirstSearch extends AStar {

	private Queue<Integer> queue;

	public BreadthFirstSearch(GridGraph graph, int sx, int sy, int ex, int ey) {
		super(graph, sx, sy, ex, ey);
		setPostSmoothingOn(false);
	}

	public static BreadthFirstSearch postSmooth(GridGraph graph, int sx, int sy, int ex, int ey) {
		BreadthFirstSearch bfs = new BreadthFirstSearch(graph, sx, sy, ex, ey);
		bfs.setPostSmoothingOn(true);
		return bfs;
	}

	@Override
	public void computePath() {
		int start = toOneDimIndex(getSx(), getSy());
		setFinish(toOneDimIndex(getEx(), getEy()));

		Arrays.fill(getParent(), 0);
		Arrays.fill(getVisited(), false);
		for (int i = 0; i < getParent().length; i++) {
			getParent()[i] = -1;
		}

		queue = new LinkedList<>();
		queue.offer(start);
		getVisited()[start] = true;

		while (!queue.isEmpty()) {
			int current = queue.poll();
			int currX = toTwoDimX(current);
			int currY = toTwoDimY(current);

			if (canGoDown(currX, currY)) {
				int index = toOneDimIndex(currX, currY - 1);
				if (!getVisited()[index]) {
					if (addToQueue(current, index))
						break;
				}
			}
			if (canGoUp(currX, currY)) {
				int index = toOneDimIndex(currX, currY + 1);
				if (!getVisited()[index]) {
					if (addToQueue(current, index))
						break;
				}
			}
			if (canGoLeft(currX, currY)) {
				int index = toOneDimIndex(currX - 1, currY);
				if (!getVisited()[index]) {
					if (addToQueue(current, index))
						break;
				}
			}
			if (canGoRight(currX, currY)) {
				int index = toOneDimIndex(currX + 1, currY);
				if (!getVisited()[index]) {
					if (addToQueue(current, index))
						break;
				}
			}

			maybeSaveSearchSnapshot();
		}

		maybePostSmooth();
	}

	/**
	 * Returns true iff finish is found.
	 */
	private boolean addToQueue(int current, int index) {
		getParent()[index] = current;
		queue.offer(index);
		getVisited()[index] = true;
		return index == getFinish();
	}

	private boolean canGoUp(int x, int y) {
		return !bottomRightOfBlockedTile(x, y) || !bottomLeftOfBlockedTile(x, y);
	}

	private boolean canGoDown(int x, int y) {
		return !topRightOfBlockedTile(x, y) || !topLeftOfBlockedTile(x, y);
	}

	private boolean canGoLeft(int x, int y) {
		return !topRightOfBlockedTile(x, y) || !bottomRightOfBlockedTile(x, y);
	}

	private boolean canGoRight(int x, int y) {
		return !topLeftOfBlockedTile(x, y) || !bottomLeftOfBlockedTile(x, y);
	}

	private boolean topRightOfBlockedTile(int x, int y) {
		return getGraph().isBlocked(x - 1, y - 1);
	}

	private boolean topLeftOfBlockedTile(int x, int y) {
		return getGraph().isBlocked(x, y - 1);
	}

	private boolean bottomRightOfBlockedTile(int x, int y) {
		return getGraph().isBlocked(x - 1, y);
	}

	private boolean bottomLeftOfBlockedTile(int x, int y) {
		return getGraph().isBlocked(x, y);
	}

}

File: library/src/main/kotlin/com/github/ohohcakester/algorithms/astar/BreadthFirstSearch.kt
package com.github.ohohcakester.algorithms.astar

import com.github.ohohcakester.grid.GridGraph

import java.util.Arrays
import java.util.LinkedList
import java.util.Queue

open class BreadthFirstSearch(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int) : AStar(graph, sx, sy, ex, ey) {
	companion object {
		fun postSmooth(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int) = postSmooth(graph, sx, sy, ex, ey, ::BreadthFirstSearch)
		fun repeatedPostSmooth(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int) = repeatedPostSmooth(graph, sx, sy, ex, ey, ::BreadthFirstSearch)
		fun dijkstra(graph: GridGraph, sx: Int, sy: Int, ex: Int, ey: Int) = dijkstra(graph, sx, sy, ex, ey, ::BreadthFirstSearch)
	}

	private lateinit var queue: Queue<Int>

	override fun computePath() {
		val start = toOneDimIndex(sx, sy)
		finish = toOneDimIndex(ex, ey)

		parent.fill(0)
		visited.fill(false)
		for (i in 0..parent.size - 1) {
			parent[i] = -1
		}

		queue = LinkedList<Int>()
		queue.offer(start)
		visited[start] = true

		while (!queue.isEmpty()) {
			val current = queue.poll()
			val currX = toTwoDimX(current)
			val currY = toTwoDimY(current)

			if (canGoDown(currX, currY)) {
				val index = toOneDimIndex(currX, currY - 1)
				if (!visited[index]) {
					if (addToQueue(current, index))
						break
				}
			}
			if (canGoUp(currX, currY)) {
				val index = toOneDimIndex(currX, currY + 1)
				if (!visited[index]) {
					if (addToQueue(current, index))
						break
				}
			}
			if (canGoLeft(currX, currY)) {
				val index = toOneDimIndex(currX - 1, currY)
				if (!visited[index]) {
					if (addToQueue(current, index))
						break
				}
			}
			if (canGoRight(currX, currY)) {
				val index = toOneDimIndex(currX + 1, currY)
				if (!visited[index]) {
					if (addToQueue(current, index))
						break
				}
			}

			maybeSaveSearchSnapshot()
		}

		maybePostSmooth()
	}

	/**
	 * Returns true iff finish is found.
	 */
	private fun addToQueue(current: Int, index: Int): Boolean {
		parent[index] = current
		queue.offer(index)
		visited[index] = true
		return index == finish
	}

	private fun canGoUp(x: Int, y: Int) = !bottomRightOfBlockedTile(x, y) || !bottomLeftOfBlockedTile(x, y)

	private fun canGoDown(x: Int, y: Int) = !topRightOfBlockedTile(x, y) || !topLeftOfBlockedTile(x, y)

	private fun canGoLeft(x: Int, y: Int) = !topRightOfBlockedTile(x, y) || !bottomRightOfBlockedTile(x, y)

	private fun canGoRight(x: Int, y: Int) = !topLeftOfBlockedTile(x, y) || !bottomLeftOfBlockedTile(x, y)

	private fun topRightOfBlockedTile(x: Int, y: Int) = graph.isBlocked(x - 1, y - 1)

	private fun topLeftOfBlockedTile(x: Int, y: Int) = graph.isBlocked(x, y - 1)

	private fun bottomRightOfBlockedTile(x: Int, y: Int) = graph.isBlocked(x - 1, y)

	private fun bottomLeftOfBlockedTile(x: Int, y: Int) = graph.isBlocked(x, y)
}
